docs/gossip.md

```diff
@@ -23,4 +24,82 @@ If you are using [bastion hosts](bastion.md), it is a bit tricky to find the bas
 
 ```
 kops toolbox dump -ojson | grep 'bastion.*elb.amazonaws.com'
-```
\ No newline at end of file
+```
+
+## Switching to memberlist from weaveworks mesh
+
+Weaveworks mesh is known to spike exponentially with CPU, memory and network bandwidth usage on cluster scale up. [issue 7429](https://github.com/kubernetes/kops/issues/7427), [issue 7436](https://github.com/kubernetes/kops/issues/7436), [issue 13974](https://github.com/kubernetes/kops/issues/13974)
+
+### Turn on memberlist as secondary protocol
+
+Add `gossipConfig` to your cluster configuration:
+
+```bash
+kops edit cluster
+```
+
+```yaml
+  gossipConfig:
+    listen: 0.0.0.0:3999
+    protocol: mesh
+    secondary:
+      listen: 0.0.0.0:4000
+      protocol: memberlist
+
+  dnsControllerGossipConfig:
+    protocol: mesh
+    listen: 0.0.0.0:3998
+    seed: 127.0.0.1:3999
+    secondary:
+      protocol: memberlist
+      listen: 0.0.0.0:3993
+```
+
+Perform a rolling-update of control-plane nodes only.
+
+Check that dns-controller is restarted and now uses the new gossip protocol (you can see that under args).
+
+```bash
+kubectl get pod -n kube-system | grep dns-controller
+kubectl get pod -n kube-system | grep dns-controller | awk '{print $1}' | xargs kubectl get pod -nkube-system -oyaml
+```
+
+```
+  - args:
+    - --watch-ingress=false
+    - --dns=gossip
+    - --gossip-protocol=mesh
+    - --gossip-listen=0.0.0.0:3998
+    - --gossip-seed=127.0.0.1:3999
+    - --gossip-protocol-secondary=memberlist
+    - --gossip-seed-secondary=127.0.0.1:3993
+    - --zone=staging.stats.superbet.com
+    - --internal-ipv4
+    - --zone=*/*
+    - -v=2
+```
+
+### 2 Switch memberlist to primary and remove secondary protocol
+
+```yaml
+  gossipConfig:
+    protocol: memberlist
+    listen: 0.0.0.0:4000
+    secondary:
+      protocol: ""
+
+  dnsControllerGossipConfig:
+    protocol: memberlist
+    listen: 0.0.0.0:3993
+    seed: 127.0.0.1:4000
+    secondary:
+      protocol: ""
+```
+
+Restart worker nodes and then restart control-plane nodes
+
+Note that this will not flag nodes as *Needs update*; you will have to do *rolling-update* with `--force` flag.
+
+**!Restart control-plane nodes last, othwerwise all nodes that still have mesh as primary will get into NotReady state!**
```

Comment:
This could be designed as three steps: add secondary, flip primary and secondary, then remove secondary. This would remove the failure if control plane updates before worker nodes.

Reply:
Original idea was:
1st add second protocol -> rolling update control planes
2nd switch protocols -> rolling update all nodes
3rd remove second protocol

but on the second step we had the the problem after rolling-update control planes as all the worker nodes went into not ready state. If we are going to do it that way, I guess we would need to do rolling-update of all nodes 3 times (on every step). This way it is just once (actually twice control planes, and once worker nodes)

I am fine with putting either in docs. We were doing switch on one of our production clusters with the two step process as doing rolling-update of all nodes three times would take us just too much time.